Python/Product/PythonTools/PythonTools/LanguageServerClient/PythonLanguageClientContextWorkspace.cs

```diff
@@ -0,0 +1,81 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABILITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.PythonTools.Common.Infrastructure;
+using Microsoft.PythonTools.Interpreter;
+using Microsoft.PythonTools.Project;
+
+namespace Microsoft.PythonTools.LanguageServerClient {
+    internal class PythonLanguageClientContextWorkspace : IPythonLanguageClientContext, IDisposable {
+        private readonly IPythonWorkspaceContext _pythonWorkspace;
+        private readonly string _contentTypeName;
+        private readonly DisposableBag _disposables;
+
+        public event EventHandler InterpreterChanged;
+        public event EventHandler SearchPathsChanged;
+        public event EventHandler Closed;
+
+        public PythonLanguageClientContextWorkspace(
+            IPythonWorkspaceContext pythonWorkspace,
+            string contentTypeName
+        ) {
+            _pythonWorkspace = pythonWorkspace ?? throw new ArgumentNullException(nameof(pythonWorkspace));
```

Comment:
Its ok to hold on to a reference to the workspace? what happens if the workspace changes? is it the same object just the internals change?

Reply:
When the workspace changes, this object will get disposed. This is kicked off by the `_pythonWorkspace.AddActionOnClose` that is a few lines below. The language client registers to this object's `Closed` event, it will terminate the language client and dispose of it, and the dispose of language client disposes of the context (this object).

Reply:
After that's all done, the events for the new workspace come in, where a new `PythonLanguageClientContextWorkspace` gets created.